source/_drafts/Feature-Spotlight-Middleware.md

```diff
@@ -0,0 +1,185 @@
+The tech is Video.js's abstraction from the player. Separating the player API from the playback technology. With techs, we can plug things like a Flash fallback or a Youtube embed into Video.js
+
+A lot of Video.js users may be familiar with middleware from projects like Express.
+Video.js middleware isn't that different from those.
+In both cases you register your middleware against a particular route to call down the chain when the route is triggered.
+In Express routes are based on the url paths.
+In Video.js, these routes are based on the video mimetype.
+And like Express, there are "star" (`*`) middleware that match all routes.
+
+There are two important pieces to be aware off with middleware:
+1. dynamic source handling,
+2. intercepting the player and tech interactions.
+
+With the dynamic source handling, you could load video with a custom type and source and resolve it asynchronously.
+A good example for this is a video catalog system.
+The page can be rendered with a specific catalog id and a special mimetype, like so:
+```html
+<video controls class="video-js">
+  <source src="123" type="video/my-catalog">
+</video>
+```
+Then, you could register a middleware for that route -- the type `video/my-catalog`.
+```js
+// middleware methods get the player instance as an argument
+videojs.use('video/my-catalog', funciton(player) {
+
+  // middleware are expected to return an object with the methods on it.
+  // It can be a plain object or an instance of something.
+  return {
+
+    // setSource allows you to tell videojs whether you're going to be handling the source or not
+    setSource(srcObj, next) {
+      const id = srcObj.src;
+
+      videojs.xhr({
+        uri: '/getVideo?id=' + id
+      }, function(err, res, body) {
+
+        // pass null as the first argument to say everything is going fine and we can handle it.
+        next(null, {
+          src: body.sourceUrl,
+          type: body.sourceType
+        })
+      });
+    }
+  };
+});
+```
+Then when Video.js initializes, it'll go through and call the middleware that are set for `video/my-catalog`.
+
+Server Side Ad Insertion (SSAI) is a great fit for middleware. It showcases the ability to intercept the play and tech interactions.
+For example, you have both ads and content in your HLS playlist but you want the controls to have a different times for each.
+When the content video is playing you want to display a timeline of 5 minutes. And for ads, you want to display 30 seconds for the pre-roll.
+Yet, currently, the timeline is showing a combined duration of 5 minutes and 30 seconds.
+So, you can add a middleware that knows when the ad is being played and tells the player that the duration is 30 seconds
+and when the content is playing that the duration is 5 minutes.
+```js
+// register a star-middleware because HLS has two mimetypes
+videojs.use('*', function(player) {
```

Comment:
Should adding an HLS middleware to star be encouraged over adding it for both mime types specifically?